### PR TITLE
Add a default value for precision field if only scale is present

### LINKED ISSUE #1300: Number scale not persisting
## Reproduce

1. In the front end, edit the DB options for a Number column.
1. Set "Number Type" to "Decimal". Set "Decimal Places" to "3". Click Save.
1. Observe a PATCH request to the columns API which specifies `type_options.scale` to `2`. Good.
1. In the response of this request (and also in the response of the subsequent GET request to the columns endpoint), expect the column to show `type_options` with an object containing a `scale` of `2`.
1. Instead, observe `type_options` to be `null`.

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #1300 

Assigns a default value of `1000` for the `precision` type_option if only `scale` type_option is sent as a request.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
@seancolsen Can you review it?

File: mathesar/api/db/viewsets/columns.py [hunk 1]
```diff
@@ -34,7 +34,8 @@ def create(self, request, table_pk=None):
         # We only support adding a single column through the API.
         serializer = ColumnSerializer(data=request.data, context={'request': request})
         serializer.is_valid(raise_exception=True)
-
+        if request.data.get('type_options', None) is not None and request.data['type_options'].get('scale', None) is not None and request.data['type_options'].get('precision', None) is None:
```
Comment (kgodey) [inline mathesar/api/db/viewsets/columns.py:37]: This is long and hard to parse quickly, can you put each conditional into a separate variable?

**Review by seancolsen — CHANGES_REQUESTED** [review]
@silentninja This is still not working for me using the exact repro steps from the ticket.

- ❌ When I send

    ```json5
    {
      "type_options": { "scale": 2 }
    }
    ```

    I expect:

    ```json5
    {
      // ...
      "type_options": { "scale": 2, "precision": 1000 }
      // ...
    }
    ```

    Instead, I receive:

    ```json5
    {
      // ...
      "type_options": null
      // ...
    }
    ```

- ❌ When I send

    ```json5
    {
      "type_options": { "scale": null, "precision": null }
    }
    ```

    I expect either of the following:

    ```json5
    {
      // ...
      "type_options": { "scale": null, "precision": null }
      // ...
    }
    ```

    ```json5
    {
      // ...
      "type_options": null
      // ...
    }
    ```

    Instead, I receive an error response:

    ```json5
    [
      {
        "code": 2023,
        "field": "precision",
        "message": "This field may not be null.",
        "detail": {}
      },
      {
        "code": 2023,
        "field": "scale",
        "message": "This field may not be null.",
        "detail": {}
      }
    ]
    ```

**Comment by silentninja** [comment]
@seancolsen 

- The last commit should fix the first error. 
-As for the second error, the backend response won't contain the fields `precision` or `scale` if their value is `None` and this being done intentionally.
https://github.com/centerofci/mathesar/blob/e33b5eed35882645f0344959fcc3f7e6c77662b6/db/columns/base.py#L180

I think the `None` value type_options are removed in order to remove irrelevant `type_options` as the backend is unable to differentiate `type_options` applicable for a type, so if I allow `null`, `type_options` will leak between type(`boolean` type will return {'fields': None, 'length': None, precision': None, 'scale': None}

The proper solution would be, to have `type` aware `type_options` which isn't a small refactor and can take quite some time.

So we have two temporary options here
- Accept `type_options` leaking into other types
- Frontend should avoid passing null values and instead remove them

**Review by seancolsen — APPROVED** [review]

**Comment by seancolsen** [comment]
Thanks @silentninja. This is looking good now.

> Frontend should avoid passing null values and instead remove them

That's fine for now. At some point it would be nice to do the refactoring you mention, so the API can better follow the principle of least surprise. But this approach works well in the short term

I've opened #1367 to track a similar problem I uncovered after I was able to test more.
